### PR TITLE
Handle service exceptions when creating organizations

### PR DESCRIPTION
**Pull request by Donnype**
## Changes
Handle service exceptions when creating organizations since this now results in crashes in Rocky.

## Issue ticket number and link
https://github.com/minvws/nl-kat-coordination/issues/702

## Proof
image

## Checklist for author(s):
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR comes from a `feature` or `hotfix` branch, in line with our git branching strategy;
- [x] This PR is "bite-sized" and only focuses on a single issue, problem, or feature;
- [x] I am not reinventing the wheel: there is no high-quality library that already has this feature;
- [x] I have changed the example `.env` files if I added, removed, or changed any config options, and I have informed others that they need to modify their `.env` files if required;
- [x] I have performed a self-review of my own code;
- [x] I have commented my code, particularly in hard-to-understand areas;
- [x] I have made corresponding changes to the documentation, if necessary;
- [x] I have written unit, integration, and end-to-end tests for the change that I made;

If a non-trivial PR:
- [ ] This PR is part of a milestone and has appropriate labels;
- [ ] This PR is properly linked to the project board (either directly or via an issue);
- [ ] I have added screenshots or some other proof that my code does what it is supposed to do;

```
## Checklist for functional reviewer(s):
- [ ] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:
* _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:
* _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
```

```
## Checklist for code reviewer(s):
- [ ] The code passes the CI tests and linters;
- [ ] The code does not bypass authentication or security mechanisms;
- [ ] The code does not introduce any dependency on a library that has not been properly vetted;
- [ ] The code does not violate Model-View-Template and our other architectural principles;
- [ ] The code contains docstrings, comments, and documentation where needed;
- [ ] The code prioritizes readability over performance where appropriate;
- [ ] The code conforms to our agreed coding standards.
```

### CODE REVIEW

File: rocky/rocky/views/organization_add.py [hunk 1]
```diff
@@ -33,13 +39,30 @@ def get_context_data(self, **kwargs):
...
     def form_valid(self, form):
-        self.object = form.save()
-        member, _ = OrganizationMember.objects.get_or_create(user=self.request.user, organization=self.object)
-        member.acknowledged_clearance_level = 4
-        member.trusted_clearance_level = 4
-        member.save()
+        try:
+            self.object = form.save()
+        except (OctopoesException, KatalogusException) as e:
+            service = "Octopoes" if isinstance(e, OctopoesException) else "the Katalogus"
```
Comment (underdarknl) [inline rocky/rocky/views/organization_add.py:45]: Woudn't it be more logical to have the service available as a value on the exception class?

Comment (ammar92) [reply rocky/rocky/views/organization_add.py:45]: Yes or:
```python3
except OctopoesException:
    # show specific error on octopoes
except KatalogusException:
    # show specific error on KATalogus
else:
    # no exceptions, redirect
```

File: rocky/rocky/views/organization_add.py [hunk 2]
```diff
@@ -33,13 +39,30 @@ def get_context_data(self, **kwargs):
...
+            message = f"An issue occurred in {service} while creating the organization"
+            logger.exception(message)
+            messages.add_message(self.request, messages.ERROR, _(message))
+
+            return redirect(self.success_url)  # get_success_url() assumes self.object is set, see ModelFormMixin
+
+        try:
+            member, created = OrganizationMember.objects.get_or_create(user=self.request.user, organization=self.object)
```
Comment (ammar92) [inline rocky/rocky/views/organization_add.py:53]: `created` is unused
```suggestion
            member, _ = OrganizationMember.objects.get_or_create(user=self.request.user, organization=self.object)
```

**Review by ammar92 — COMMENTED** [review]
A few small remarks. Tests look great 👍 One general remark about naming: since KAT is an abbreviation and used as such in both our UI as codebase, you should also be consistent when naming the classes, e.g.: `KATalogusException` instead of `KatalogusException`

File: rocky/rocky/views/organization_add.py (hunk 1 above)
Comment (Donnype) [reply rocky/rocky/views/organization_add.py:45]: I also considered this, but was a bit hesitant as that would add representation data in the exception.

Comment (Donnype) [reply rocky/rocky/views/organization_add.py]: I added a base class called `ServiceException` to represent issues with (external) services and carries the name.

**Review by ammar92 — APPROVED** [review]

**Review by underdarknl — APPROVED** [review]

**Pull request merged**